src/js_notifications.hpp

```diff
@@ -0,0 +1,163 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+#include <map>
+#include <vector>
+
+#include <realm/object-store/collection_notifications.hpp>
+
+#include "js_types.hpp"
+
+namespace realm::js::notifications {
+
+using IdType = unsigned long long;
+
+template <typename T>
+class NotificationHandle;
+
+/**
+ * @brief A class with static members use to manage ownership of `NotificationToken`s returned by the object store
+ * notification APIs.
+ * This abstraction is needed to enable graceful JS runtime destruction by preventing circular references from objects
+ * owning a `NotificationToken` owning a lamda that captures `Protected` values of the object itself.
+ *
+ * @note This expose a simple `clear` method, which should be called just before the JS runtime is torn down.
+ *
+ * @tparam T The JS runtime types.
+ */
+template <typename T>
+class NotificationBucket {
+    using ProtectedFunction = Protected<typename T::Function>;
+    using TokensMap = std::map<IdType, std::vector<std::pair<ProtectedFunction, NotificationToken>>>;
+
+public:
+    static void emplace(NotificationHandle<T>& handle, ProtectedFunction&& callback, NotificationToken&& token)
+    {
+        if (handle) {
+            auto& s_tokens = get_tokens();
+            s_tokens[handle].emplace_back(std::move(callback), std::move(token));
+        }
+        else {
+            throw std::runtime_error("Cannot emplace notifications using an unset handle");
+        }
+    }
+
+    static void clear()
+    {
+        auto& s_tokens = get_tokens();
+        s_tokens.clear();
+    }
+
+    static void erase(NotificationHandle<T>& handle)
+    {
+        if (handle) {
+            auto& s_tokens = get_tokens();
+            s_tokens.erase(handle);
+        }
+    }
+
+    static void erase(NotificationHandle<T>& handle, ProtectedFunction&& callback)
+    {
+        if (handle) {
+            auto& s_tokens = get_tokens();
+            auto& tokens = s_tokens[handle];
+            auto compare = [&callback](auto&& token) {
+                return typename ProtectedFunction::Comparator()(token.first, callback);
+            };
+            tokens.erase(std::remove_if(tokens.begin(), tokens.end(), compare), tokens.end());
+        }
+        else {
+            throw std::runtime_error("Cannot erase notifications using an unset handle");
+        }
+    }
+
+    /**
+     * @brief Get the static tokens object.
+     *
+     * @note We cannot use a simple static object on the namespace, because we are referencing this across multiple
+     * translation units and this could lead to code using this before it's initialized.
+     *
+     * @return auto& A reference to the static map of tokens.
+     */
+    inline static TokensMap& get_tokens()
+    {
+        static TokensMap s_tokens;
```

Comment:
[question] I just am curious what `s_` indicates? (Probably something about it being static). I would also like to know how this does get initialised.  Does static in this case behave a bit like a singleton?

Reply:
> Probably something about it being static.

Right. I personally don't like this convention, but I believe it got introduced in the previous implementation because other pieces of the code adhered to it.

> I would also like to know how this does get initialised.

This is the initialisation - this syntax calls the default constructor with no arguments.